src/voice/udp/handle.rs

```diff
@@ -0,0 +1,256 @@
+
+        self.send_rtp_packet(rtp_packet).await
+    }
+
+    /// Encrypts and sends and rtp packet.
+    ///
+    /// # Errors
+    /// If we have not received an encryption key, this returns a [VoiceUdpError::NoKey] error.
+    ///
+    /// If the Udp socket is broken, this returns a [VoiceUdpError::BrokenSocket] error.
+    pub async fn send_rtp_packet(
+        &self,
+        packet: discortp::rtp::MutableRtpPacket<'_>,
+    ) -> Result<(), VoiceUdpError> {
+        let mut buffer = self.encrypt_rtp_packet_payload(&packet).await?;
+        let new_packet = discortp::rtp::MutableRtpPacket::new(&mut buffer).unwrap();
+        self.send_encrypted_rtp_packet(new_packet.consume_to_immutable())
+            .await?;
+        Ok(())
+    }
+
+    /// Encrypts an unencrypted rtp packet, returning a copy of the packet's bytes with an
+    /// encrypted payload
+    ///
+    /// # Errors
+    /// If we have not received an encryption key, this returns a [VoiceUdpError::NoKey] error.
+    ///
+    /// When using voice encryption modes which require special nonce generation, and said generation fails, this returns a [VoiceUdpError::FailedNonceGeneration] error.
+    pub async fn encrypt_rtp_packet_payload(
+        &self,
+        packet: &discortp::rtp::MutableRtpPacket<'_>,
+    ) -> Result<Vec<u8>, VoiceUdpError> {
+        let payload = packet.payload();
+
+        let session_description_result = self.data.read().await.session_description.clone();
+
+        // We are trying to encrypt, but have not received SessionDescription yet,
+        // which contains the secret key.
+        if session_description_result.is_none() {
+            return Err(VoiceUdpError::NoKey);
+        }
+
+        let session_description = session_description_result.unwrap();
+
+        let mut nonce_bytes = match session_description.encryption_mode {
+            VoiceEncryptionMode::Xsalsa20Poly1305 => get_xsalsa20_poly1305_nonce(packet.packet()),
+            VoiceEncryptionMode::Xsalsa20Poly1305Suffix => {
+                // Generate 24 random bytes
+                let mut random_destinaton: Vec<u8> = vec![0; 24];
+                let random_result = getrandom(&mut random_destinaton);
+                if let Err(e) = random_result {
+                    return Err(VoiceUdpError::FailedNonceGeneration {
+                        error: format!("{:?}", e),
+                    });
+                }
+                random_destinaton
+            }
+            VoiceEncryptionMode::Xsalsa20Poly1305Lite => {
+                // "Incremental 4 bytes (32bit) int value"
+                let mut data_lock = self.data.write().await;
+                let nonce = data_lock
+                    .last_udp_encryption_nonce
+                    .unwrap_or_default()
+                    .wrapping_add(1);
+
+                data_lock.last_udp_encryption_nonce = Some(nonce);
+                drop(data_lock);
+                // TODO: Is le correct? This is not documented anywhere
```

Comment:
Network byte order is usually BE. No clue about this specific case, though

Reply:
True, I've changed it to big endian